internal/extsvc/github/common.go

```diff
@@ -1517,7 +1517,9 @@ func doRequest(ctx context.Context, logger log.Logger, apiURL *url.URL, auther a
 		autherWithRefresh, ok = auther.(auth.AuthenticatorWithRefresh)
 		// Check if we should pre-emptively refresh
 		if ok && autherWithRefresh.NeedsRefresh() {
-			autherWithRefresh.Refresh(ctx, httpClient)
+			if err := autherWithRefresh.Refresh(ctx, httpClient); err != nil {
+				logger.Warn("doRequest: refresh token failed", log.Error(err))
```

Comment:
warn logs won't end up in Sentry. I think we can add a method like `NeedsRefresh` which checks whether the token is _already expired_ and in this case log with `error` level which will create a Sentry event, which is crucial in such situation IMO.

WDYT?

Reply:
`oauthutil.DoRequest` already does it here a little further below: https://github.com/sourcegraph/sourcegraph/blob/fd67e1906b13a30badf4d4eb340e083776f25184/internal/extsvc/github/common.go#L1541-L1545

So if the token is expired we will end up refreshing it and making the request. I don't see much value in implementing a check for already expired token in this case.

Reply:
Since we hard fail from `DoRequest`, I think it's fine to ignore the error the first time especially if we cannot be certain if the error is intermittent or permanent. My thinking is if this happens frequently then we should see the errors propagate out from the above code and it won't be ignored.

Reply:
I think a better phrasing here would be
```suggestion
				logger.Warn("doRequest: refreshing of the token failed", log.Error(err))
```

Because `'refresh token'` might refer to the refresh token itself, not the action of refreshing the token.
Should I create a follow up PR or do you want to do it yourself @indradhanush ?

Reply:
Good point. PR here: https://github.com/sourcegraph/sourcegraph/pull/44068